### ISSUE TITLE
clustermesh: downtime/dropped packets due to network policy on upgrade

### ISSUE THREAD

**Issue by rcanderson23** [cilium/cilium#26462]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

We are currently running Cilium on EKS in chaining mode with aws-cni. We are utilizing clustermesh across 6 clusters and the configuration and installation is managed by FluxCD. Our application has an auth service that lives in one of the six clusters and proxy service that lives in all clusters. The proxies connect to each other in a mesh as well as initiate a connection to the auth service. The proxies rely on a global service to connect to auth, they dial each other directly via a different discovery mechanism.

Single cluster upgrades don't appear to be an issue but if multiple clusters upgrade close enough together we observe connection issues between proxies in those clusters. Specifically we are looking at the metric `hubble_drop_total{reason="Policy denied"}` as well as metric on our application that indicates the number of other proxies a proxy is connected to. Looking at the logs of the cilium agent of the node where affected proxies were running it seems like it isn't uncommon to see a connection to etcd, disconnect, and reconnect. It is fairly reproducible by running `kubectl rollout restart deployment clustermesh-apiserver` in cluster1 and `kubectl rollout restart ds cilium` in cluster2.

I considered trying to split the cilium install and the clustermesh install into separate helm installs as that would give us control when they would restart and not happen at the same time. Unfortunately, the `cilium-ca` secret is expected to be created every time and would cause a conflict (there may be other but this was the first roadblock I hit, could potentially get around this by allowing users to specify this).

We are hoping to better understand the following:
1. Expectations of upgrades as the upgrade guide states `Minimal to None` affect on L3/L4 and if we are hitting that or if we are hitting something else.
2. Is it expected that Cilium upgrades in clustermesh happen one cluster at a time? Our FluxCD reconciles every 10 min and currently have no coordination that at the moment but are looking at options for this.
3. Would we be better off configuring our cilium install to be backed by etcd? That seems like it would remove the possibility of etcd connections being broken on cilium startup since there wouldn't be a `clustermesh-apiserver` pod being recreated. Is there documentation for this by chance?
4. Possibly separate the `etcd` container (and give it a PV) from the `apiserver` container so that clustermesh isn't rolled at the same time? Any gotchas with this approach?

Thanks for all the work on this awesome CNI!

### Cilium Version

1.12.3

### Kernel Version

Linux ip-10-16-112-106.us-west-2.compute.internal 5.4.242-155.348.amzn2.x86_64 #1 SMP Mon May 8 12:52:40 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

Server Version: v1.23.17-eks-0a21954

### Sysdump

Sysdump file was too large to upload

### Relevant log output

```shell
level=info msg="Established connection to remote etcd" clusterName=redact config=/var/lib/cilium/clustermesh/redact kvstoreErr="<nil>" kvstoreStatus="etcd: 1/1 connected, lease-ID=0, lock lease-ID=0, has-quorum=timeout while waiting for initial connection, consecutive-errors=1: https://redact:2379 - 3.5.4 (Leader)" subsys=clustermesh
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by rcanderson23** [comment]
I did some additional testing here. 

Observing the hubble UI during my testing I see my traffic generator app being classified as `Unknown app` and `world` until eventually it gets the correct identity and traffic starts being allowed once more.

v1.14.0-rc.0:
Using `kvstoremesh` I saw roughly 5 minutes dropped packets with `POLICY_DENIED` as the reason when cilium agents and the clustermesh-apiserver pod were restarted in the same cluster. Without `kvstoremesh` enabled and using the original test method of restarting cilium agents in one cluster and clustermesh-apiserver in another I saw between 60s and 120s of dropped packets due to policy enforcement.

v1.12.3->v1.12.11->v1.13.4->v1.14.0-rc.0 using `etcd`:
0 or near 0 increments in `hubble_drop_total{reason="Policy denied"}` so I am guessing this the best path forward for Cilium upgrades and avoiding packet drops due to policy enforcement.

**Comment by giorio94** [comment]
Thanks @rcanderson23 for the detailed issue!

I've been able to consistently reproduce the connectivity issue scaling the clustermesh-apiserver to zero replicas in one cluster and restarting the cilium agents in the other. Essentially, the issue happens when, during the bootstrap phase, the agents regenerate all local endpoints, switching, among the others, to using the newly populated ipcache map. Yet, at that point, the ipcache map will correctly contain all local entries, while the ones referring to the remote clusters might still be missing, breaking the connectivity towards remote endpoints. Similarly, policy entries get also synchronized with the known state, possibly removing entries referring to remote endpoints if not yet discovered. 

Please note that scaling the clustermesh-apiserver to zero replicas makes this issue easy to reproduce, but it would also happen restarting the agents only, if the endpoints get regenerated before synchronizing all ipcache entries from the remote etcd --- it would just be less likely (depending on the total number of remote endpoints.

It is not yet fully clear to me a possible solution to limit (or better fix) this issue. The tricky part about solving this issue is not introduce a hard dependency on remote clusters, as that could easily lead to chicken-and-egg issues preventing the entire cluster mesh from converging.

--------

> Looking at the logs of the cilium agent of the node where affected proxies were running it seems like it isn't uncommon to see a connection to etcd, disconnect, and reconnect.

There's one bug I'm aware of (and that has been recently fixed by https://github.com/cilium/cilium/pull/26363) which caused a valid connection to the remote kvstore to be restarted, typically happening when the clustermesh-apiserver and the agents get restarted at the same time. I guess that it is what you observed in this case.

> Expectations of upgrades as the upgrade guide states Minimal to None affect on L3/L4 and if we are hitting that or if we are hitting something else.

My understanding is that the upgrade guide expectations currently apply to single cluster scenarios, while there are still a few rough edges that would benefit from improvements in the multi cluster case (a few known issues have also already been fixed in v1.14). 

> Using kvstoremesh I saw roughly 5 minutes dropped packets with POLICY_DENIED as the reason when cilium agents and the clustermesh-apiserver pod were restarted in the same cluster. Without kvstoremesh enabled and using the original test method of restarting cilium agents in one cluster and clustermesh-apiserver in another I saw between 60s and 120s of dropped packets due to policy enforcement.

It is expected that kvstoremesh introduces a bit of additional latency, but this difference seems pretty severe to me. Could you please share the logs of the cilium agents, clustermesh-apiserver and kvstoremesh containers from that run to double-check if anything wrong happened?

**Comment by rcanderson23** [comment]
Sorry for the late reply and thanks for the detailed answer. This definitely helps us determine our best path forward. 

> It is expected that kvstoremesh introduces a bit of additional latency, but this difference seems pretty severe to me. Could you please share the logs of the cilium agents, clustermesh-apiserver and kvstoremesh containers from that run to double-check if anything wrong happened?

Unfortunately, it was with an ephemeral setup and don't have logs. I retested and my results this time seem to show similar results in a 3 node cluster with 50 "workload namespaces with global services" with kvstoremesh enabled and disabled so it is possible something abnormal happened that one time as I don't seem to be able to repeat it.

**Comment by giorio94** [comment]
@rcanderson23 The fixes to prevent the connection drops mentioned in this issue have just been merged (and are being backported to v1.14). Please feel free to reopen this issue (or a new one) if you still encounter problems.
